flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkChangelogModeInferenceProgram.scala

```diff
@@ -220,6 +221,38 @@ class FlinkChangelogModeInferenceProgram extends FlinkOptimizeProgram[StreamOpti
         val providedTrait = ModifyKindSetTrait.INSERT_ONLY
         createNewNode(rel, children, providedTrait, requiredTrait, requester)
 
+      case rank: StreamPhysicalRank if RankUtil.isDeduplication(rank) =>
+        val children = visitChildren(rel, ModifyKindSetTrait.ALL_CHANGES)
+        val tableConfig = unwrapTableConfig(rank)
+
+        // if the rank is deduplication and can be executed as insert-only, forward that information
+        val insertOnly = children
+          .filterNot(
+            rel => {
+              rel.getTraitSet.contains(ModifyKindSetTrait.INSERT_ONLY)
+            })
+          .isEmpty
+
+        val providedTrait = {
+          if (
+            insertOnly && StreamExecDeduplicate.canBeInsertOnly(
```

Comment:
It's a bit strange to use the exec node method in a place that only handles physical nodes. How about moving this method to `RankUtil` and naming it something like `RankUtil#outputInsertOnlyInDeduplicate`?